### PR TITLE
Fixed testhost crash for net7

### PR DESCRIPTION
**Pull request by cvpoienaru**
This PR fixes two issues:
- Testhost crashing when using the in-proc wrapper to run net7 tests because dotnet couldn't correctly resolve the net7 runtime. Because we now run in-proc (meaning inside the `TestWindowStoreHost` process), certain environment variables such as `DOTNET_ROOT` and `DOTNET_MULTILEVEL_LOOKUP` are set. We cannot unset them because that would change the behavior of the `TestWindowStoreHost` process, but we certainly need them unset for dotnet runtime resolution to be successful. AppDomains provide no isolation since environment variables are shared at the process level. The simplest and most elegant solution is to set the `VSTEST_WINAPPHOST_DOTNET_ROOT` and this would provide successful resolution. More info about `VSTEST_WINAPPHOST_DOTNET_ROOT` can be found [here](https://github.com/microsoft/vstest/pull/3715).

Error here:
```
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process for source(s) 'C:\Users\<USER>\source\repos\TestProject12\TestProject12\bin\Debug\net7.0\TestProject12.dll' exited with error: You must install or update .NET to run this application.
App: C:\Users\<USER>\source\repos\TestProject12\TestProject12\bin\Debug\net7.0\testhost.exe
Architecture: x64
Framework: 'Microsoft.NETCore.App', version '7.0.0-rc.2.22472.3' (x64)
.NET location: C:\Program Files\Microsoft Visual Studio\2022\Preview\dotnet\runtime
The following frameworks were found:
  6.0.9 at [C:\Program Files\Microsoft Visual Studio\2022\Preview\dotnet\runtime\shared\Microsoft.NETCore.App]
Learn about framework resolution:
https://aka.ms/dotnet/app-launch-failed
To install missing framework, download:
https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=7.0.0-rc.2.22472.3&arch=x64&rid=win10-x64
. Please check the diagnostic logs for more information.
```
- An unimplemented handler for raw messages got called on TW side when running in-proc. With out-of-proc, raw messages never reach TW since they aren't dequeued by the translation layer. Duplicating the same logic here, since TW wouldn't even know what to do with raw messages. This is not indicating a malfunction on either TP or TW side, but rather the lack of adherence to a convention. Even if this exception is logged in the output window, the tests are still successfully discovered/run.

Error here:
```
System.NotImplementedException: The method or operation is not implemented.
   at Microsoft.VisualStudio.TestWindow.Host.TestPlatformProvider.TestResponseHandler.HandleRawMessageAsync(String rawMessage)
   at Microsoft.VisualStudio.TestWindow.RemoteAgent.TestAgentBridgeCallback.HandleRawMessageAsync(String rawMessage)
   at Microsoft.VisualStudio.TestWindow.Logging.ILoggerExtensions.<CallWithCatchAsync>d__10.MoveNext()
```

### CODE REVIEW

**Comment by nohwnd** [comment]
Please provide more information about the fix and the reasoning the description.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (nohwnd) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:92]: This fixes this particular problem, but makes the behavior different between using in-process console and not. When in-process console is not used, we are given a full set of env variables when wrapper is instantiated, and we use those to start vstest.console. Testhost process (and other processes started by vstest.console) then inherits just those env variables. 

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/ConsoleParameters.cs#L42-L55

In-process console should probably do the same, and use those env variables to start additional processes. Otherwise the environment will be different for each usecase.

Comment (cvpoienaru) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs:92]: Many thanks for the suggestion ! I've actually looked into this first thing when trying to figure out a way to fix the issue and I agree, it would be ideal if we could fix it like you suggested, but unfortunately it doesn't seem possible (at least from my POV).

The reason is that the in-process wrapper, by definition, runs in process, meaning in the `TestWindowStoreHost` process where there's no isolation we can use like AppDomains since environment variables are set at the process level and this seems to be the most granular option we get. We cannot clear the environment variables inside the in-process wrapper and only set the env variables we get via console parameters because that would mess with the TW process. We get the best of both worlds really by keeping the env vars inherited from the TW and (eventually) overriding them because the env var sets are mostly identical, the only thing being left intact is the env vars exclusive to TW.

Propagating the env var set from console params through TP layers down to testhost and other child processes like data collectors is not trivial and the problem we're trying to solve at this point is more theoretical than a real problem. Because of that I'd rather better use this time someplace else. Moreover, we don't really know what a testhost really is, it can be a separate process but it can be loaded as an extension in vstest.console (as with UWP) and then we get the same issue since the testhost will run as part of the vstest.console process, which in fact will be the same process as `TestWindowStoreHost`. We'd have to convince the UWP folks (and other extension writers on the way) to do this propagation when spawning their own processes if they happen to do this.

All in all, I feel confident we can go with this approach at least for now and fix other issues that may appear on the way rather than undertaking the task of propagating the env vars as outlined above, which is not a bulletproof approach either ...

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (Evangelink) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:152]: @MarcoRossignoli Can you think of a better way to retrieve the "default" global installation folder?

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (Evangelink) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:141]: This is not enough, you need to check for the x86 one too. See https://github.com/microsoft/vstest/blob/9a5bb2faf130809253b28d7ad2d434d785901025/src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs#L482-L489

**Review by Evangelink — COMMENTED** [review]
As I said during our talk, this seems to be working but I am quite worried this will end up reverted like the DOTNET_ROOT update I did because we are too far away from understanding all the scenario.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (MarcoRossignoli) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:158]: Looks to me that we're changing this env var for "current" process and so we're affecting not only our host, is it right?
Should it be done before the host start?

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (MarcoRossignoli) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs:152]: In windows can be in different way...99.9% I expect to find it there...but it's possible that user will install in different folder.
The "correct code" is here:

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs#L121

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (nohwnd) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs:92]: There is existing abstraction over environment, I think you could use that. And make sure it is used everywhere you need it.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (cvpoienaru) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs]: Removed this method altogether.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (cvpoienaru) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs]: Indeed, we altered the environment of the current process. I've removed this fix and instead propagated the env variables only to the testhost process.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (cvpoienaru) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs]: Closing this as we abandoned the idea of patching the dotnet location.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs
Comment (MarcoRossignoli) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs:48]:
```suggestion
    public static IDictionary<string, string?>? DefaultEnvironmentVariables { get; set; }
```

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (nohwnd) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:68]: Is there a good reason to add another internal constructor when there already is one?

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (nohwnd) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:123]: You are not passing the option that says if we should clean up, so if user says we should not (default) and will pass just AAA="bbb", you won't be able to start testhost, because it will get just 1 env variable.

File: test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs
Comment (nohwnd) [inline test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs:136]: This should make sure there are also the other env variables, because right now there probably is just 1, but there should be all of them, with AAAA added.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs
Comment (nohwnd) [reply src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs:48]: I like the original name better, because this is not the default.

Comment (Evangelink) [reply src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs:48]: I also prefer the original name. Maybe we could go with `ProcessEnvironmentVariables` or `BaselineEnvironmentVariables`?

File: src/Microsoft.TestPlatform.Client/AttachmentsProcessing/InProcessTestRunAttachmentsProcessingEventsHandler.cs
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.Client/AttachmentsProcessing/InProcessTestRunAttachmentsProcessingEventsHandler.cs:20]: Done.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (cvpoienaru) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs:68]: Not really, I was being lazy and I didn't want to change all the tests, but for clarity we should have only 1 internal constructor, I agree. Making the change now 👍

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs:105]: We directly pass the environment variable collection we get via `ConsoleParameters`, which has the following type: `Dictionary<string, string?>`, to the `ProcessHelper` property. Note the value in the key value pair can be null. If we don't guard against a null value here, we'll end up setting an environment variable that has null as value, which instead has the effect of deleting the said environment var from the collection at some point. Should that variable not exist in the first place, we can probably expect an exception being thrown. This is just trying to guard against that possibility.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs
Comment (cvpoienaru) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs:123]: Awesome catch, thanks ! Fixing right away 👍

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs
Comment (MarcoRossignoli) [reply src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs:48]: I have indented this prop like "you inherit from parent and plus you'll always add this default list" so "external" is not super clear to me, as a static "always present" environment variables. But I'm fine with both.

File: test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs
Comment (cvpoienaru) [reply test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs:136]: Do you mean the final set of environment variables that is used to spawn the testhost should contain, among others, `AAAAA` as well ? Because unfortunately we cannot test that, we don't have access to the process start info due to the way `ProcessHelper` is architectured. I'm just making sure the in-process wrapper correctly sets the console params env variables for the `ProcessHelper` to use, so that we don't regress.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs
Comment (cvpoienaru) [reply src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs]: I think I may have partially addressed this naming issue by fixing another comment I got from Jakub. I've added an extra property called `InheritEnvironmentVariables` to be used in tandem with `ExternalEnvironmentVariables`. If `InheritEnvironmentVariables` is set to true, then those env variables are just extra variables to be added to the ones we've already inherited from the parent process, otherwise we first clean up the baseline and add only our external environment variables afterwards.

File: src/vstest.console/vstest.console.csproj
Comment (Evangelink) [inline src/vstest.console/vstest.console.csproj:53]: Hack to work around the issue where nullable attributes are coming from both PlatformAbstractions and CoreUtilities. We have tried to solve it in a better way but that was leading to too complex refactoring. Using the alias trick helps to isolate what's coming from this assembly.

Note that I hope to remove such hack by using a source generator for the polyfill (@MarcoRossignoli do you remember the name of the project?)

File: src/vstest.console/vstest.console.csproj
Comment (Evangelink) [inline src/vstest.console/vstest.console.csproj:25]: Using this trick as some of the files that need to be fixed are included from a shared project that would force us to define the alias in all using projects.

Comment (Evangelink) [reply src/vstest.console/vstest.console.csproj:25]: @cvpoienaru Re-reading that now, I am thinking that it could be clearer to use `USE_EXTERN_ALIAS` instead, WDYT?

**Review by Evangelink — APPROVED** [review]
LGTM. Please wait for approval from @nohwnd and @MarcoRossignoli before merging (I have disabled auto-merge).

File: test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs
Comment (nohwnd) [reply test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs:140]: The test is not testing if inherit env variables is used correctly. If you pass AAAA=1, then in both cases checking AAAA=1  will be true. But you don't know anything about whether AAAA=1 is now the only env variables in place (inherit=false), or if AAAA=1 was added to the existing set of variables (inherit=true).

File: src/Microsoft.TestPlatform.Execution.Shared/DebuggerBreakpoint.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.Execution.Shared/DebuggerBreakpoint.cs:4]: This seems bit unusual, what is the reason to have it and are there other ways?

Comment (nohwnd) [reply src/Microsoft.TestPlatform.Execution.Shared/DebuggerBreakpoint.cs:4]: Answered just above.

**Comment by nohwnd** [comment]
I would prefer to not propagate the inherit flag all the way down to ProcessHelper, instead the code that sets the variables (in wrapper) should check what the preference is and either provide just what it was given (inherit=false), or merge it with the current set of env variables (inherit=true) and set that to ExternalEnvVariables. This way providing external env variables will work the same no matter if we are inheriting or not in terms of updates etc.

File: src/vstest.console/vstest.console.csproj
Comment (MarcoRossignoli) [reply src/vstest.console/vstest.console.csproj:53]: https://github.com/Sergio0694/PolySharp

**Review by nohwnd — APPROVED** [review]

File: test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs
Comment (cvpoienaru) [reply test/vstest.console.UnitTests/InProcessVsTestConsoleWrapperTests.cs]: Fixed, thanks !

**Review by MarcoRossignoli — APPROVED** [review]
Pls @cvpoienaru when you have a pair of hours can you add the explanation of the in-process working with this "details" here https://github.com/microsoft/vstest-docs/tree/main/RFCs

And maybe after link in the code here, for future devs, we're moving in different context and this repo doesn't have any information of the other world.

**Comment by cvpoienaru** [comment]
Follow up: https://github.com/microsoft/vstest/issues/4119

**Pull request merged**
